File: package/ios/Sources/AdvancedFileManagerPlugin/AdvancedFileManagerPlugin.swift
import Foundation
import Capacitor
import UIKit
import UniformTypeIdentifiers

/**
 * Advanced File Manager Plugin for Capacitor
 * Provides comprehensive file system operations
 */
@objc(AdvancedFileManagerPlugin)
public class AdvancedFileManagerPlugin: CAPPlugin, CAPBridgedPlugin {
    public let identifier = "AdvancedFileManagerPlugin"
    public let jsName = "AdvancedFileManager"
    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "requestPermissions", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "checkPermissions", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "listDirectory", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "createDirectory", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "deleteDirectory", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "createFile", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "readFile", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "writeFile", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "deleteFile", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "moveFile", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "copyFile", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "renameFile", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "getFileInfo", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "exists", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "openSystemFilePicker", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "openSystemFileManager", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "openFileWithSystemApp", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "echo", returnType: CAPPluginReturnPromise)
    ]
    private let implementation = AdvancedFileManager()

    @objc public override func requestPermissions(_ call: CAPPluginCall) {
        // iOS 应用在沙盒内有完整的文件访问权限
        call.resolve([
            "granted": true,
            "message": "File access permissions granted within app sandbox"
        ])
    }

    @objc public override func checkPermissions(_ call: CAPPluginCall) {
        call.resolve([
            "granted": true,
            "message": "File access permissions granted within app sandbox"
        ])
    }

    @objc func listDirectory(_ call: CAPPluginCall) {
        guard let path = call.getString("path") else {
            call.reject("Path is required")
            return
        }

        let showHidden = call.getBool("showHidden") ?? false
        let sortBy = call.getString("sortBy") ?? "name"
        let sortOrder = call.getString("sortOrder") ?? "asc"

        do {
            let result = try implementation.listDirectory(path: path, showHidden: showHidden, sortBy: sortBy, sortOrder: sortOrder)
            call.resolve(result)
        } catch {
            call.reject("Failed to list directory: \(error.localizedDescription)")
        }
    }

    @objc func createDirectory(_ call: CAPPluginCall) {
        guard let path = call.getString("path") else {
            call.reject("Path is required")
            return
        }

        let recursive = call.getBool("recursive") ?? false

        do {
            try implementation.createDirectory(path: path, recursive: recursive)
            call.resolve()
        } catch {
            call.reject("Failed to create directory: \(error.localizedDescription)")
        }
    }

    @objc func deleteDirectory(_ call: CAPPluginCall) {
        guard let path = call.getString("path") else {
            call.reject("Path is required")
            return
        }

        do {
            try implementation.deleteDirectory(path: path)
            call.resolve()
        } catch {
            call.reject("Failed to delete directory: \(error.localizedDescription)")
        }
    }

    @objc func createFile(_ call: CAPPluginCall) {
        guard let path = call.getString("path") else {
            call.reject("Path is required")
            return
        }

        let content = call.getString("content") ?? ""
        let encoding = call.getString("encoding") ?? "utf8"

        do {
            try implementation.createFile(path: path, content: content, encoding: encoding)
            call.resolve()
        } catch {
            call.reject("Failed to create file: \(error.localizedDescription)")
        }
    }

    @objc func readFile(_ call: CAPPluginCall) {
        guard let path = call.getString("path") else {
            call.reject("Path is required")
            return
        }

        let encoding = call.getString("encoding") ?? "utf8"

        do {
            let result = try implementation.readFile(path: path, encoding: encoding)
            call.resolve(result)
        } catch {
            call.reject("Failed to read file: \(error.localizedDescription)")
        }
    }

    @objc func writeFile(_ call: CAPPluginCall) {
        guard let path = call.getString("path"),
              let content = call.getString("content") else {
            call.reject("Path and content are required")
            return
        }

        let encoding = call.getString("encoding") ?? "utf8"
        let append = call.getBool("append") ?? false

        do {
            try implementation.writeFile(path: path, content: content, encoding: encoding, append: append)
            call.resolve()
        } catch {
            call.reject("Failed to write file: \(error.localizedDescription)")
        }
    }

    @objc func deleteFile(_ call: CAPPluginCall) {
        guard let path = call.getString("path") else {
            call.reject("Path is required")
            return
        }

        do {
            try implementation.deleteFile(path: path)
            call.resolve()
        } catch {
            call.reject("Failed to delete file: \(error.localizedDescription)")
        }
    }

    @objc func moveFile(_ call: CAPPluginCall) {
        guard let sourcePath = call.getString("sourcePath"),
              let destinationPath = call.getString("destinationPath") else {
            call.reject("Source path and destination path are required")
            return
        }

        do {
            try implementation.moveFile(sourcePath: sourcePath, destinationPath: destinationPath)
            call.resolve()
        } catch {
            call.reject("Failed to move file: \(error.localizedDescription)")
        }
    }

    @objc func copyFile(_ call: CAPPluginCall) {
        guard let sourcePath = call.getString("sourcePath"),
              let destinationPath = call.getString("destinationPath") else {
            call.reject("Source path and destination path are required")
            return
        }

        let overwrite = call.getBool("overwrite") ?? false

        do {
            try implementation.copyFile(sourcePath: sourcePath, destinationPath: destinationPath, overwrite: overwrite)
            call.resolve()
        } catch {
            call.reject("Failed to copy file: \(error.localizedDescription)")
        }
    }

    @objc func renameFile(_ call: CAPPluginCall) {
        guard let path = call.getString("path"),
              let newName = call.getString("newName") else {
            call.reject("Path and new name are required")
            return
        }

        do {
            try implementation.renameFile(path: path, newName: newName)
            call.resolve()
        } catch {
            call.reject("Failed to rename file: \(error.localizedDescription)")
        }
    }

    @objc func getFileInfo(_ call: CAPPluginCall) {
        guard let path = call.getString("path") else {
            call.reject("Path is required")
            return
        }

        do {
            let result = try implementation.getFileInfo(path: path)
            call.resolve(result)
        } catch {
            call.reject("Failed to get file info: \(error.localizedDescription)")
        }
    }

    @objc func exists(_ call: CAPPluginCall) {
        guard let path = call.getString("path") else {
            call.reject("Path is required")
            return
        }

        let exists = implementation.exists(path: path)
        call.resolve(["exists": exists])
    }

    @objc func openSystemFilePicker(_ call: CAPPluginCall) {
        let type = call.getString("type") ?? "file"
        let multiple = call.getBool("multiple") ?? false
        _ = call.getString("title") ?? "选择文件"

        DispatchQueue.main.async {
            if type == "directory" {
                // iOS 14+ 支持目录选择
                if #available(iOS 14.0, *) {
                    let documentPicker = UIDocumentPickerViewController(forOpeningContentTypes: [.folder])
                    documentPicker.allowsMultipleSelection = multiple
                    documentPicker.delegate = self

                    // 存储 call 以便在回调中使用
                    self.pendingCall = call

                    self.bridge?.viewController?.present(documentPicker, animated: true)
                } else {
                    call.reject("Directory selection requires iOS 14+")
                }
            } else {
                // 文件选择
                let documentPicker = UIDocumentPickerViewController(forOpeningContentTypes: [.item])
                documentPicker.allowsMultipleSelection = multiple
                documentPicker.delegate = self

                self.pendingCall = call

                self.bridge?.viewController?.present(documentPicker, animated: true)
            }
        }
    }

    @objc func openSystemFileManager(_ call: CAPPluginCall) {
        _ = call.getString("path")

        DispatchQueue.main.async {
            // iOS 没有独立的文件管理器应用，打开 Files 应用
            if let url = URL(string: "shareddocuments://") {
                if UIApplication.shared.canOpenURL(url) {
                    UIApplication.shared.open(url) { success in
                        if success {
                            call.resolve()
                        } else {
                            call.reject("Failed to open Files app")
                        }
                    }
                } else {
                    call.reject("Files app not available")
                }
            } else {
                call.reject("Invalid URL for Files app")
            }
        }
    }

    @objc func openFileWithSystemApp(_ call: CAPPluginCall) {
        guard let filePath = call.getString("filePath") else {
            call.reject("File path is required")
            return
        }

        let fileURL = URL(fileURLWithPath: filePath)

        DispatchQueue.main.async {
            let documentController = UIDocumentInteractionController(url: fileURL)
            documentController.delegate = self

            if let viewController = self.bridge?.viewController {
                if documentController.presentPreview(animated: true) {
                    call.resolve()
                } else if documentController.presentOpenInMenu(from: viewController.view.bounds, in: viewController.view, animated: true) {
                    call.resolve()
                } else {
                    call.reject("No app available to open this file type")
                }
            } else {
                call.reject("View controller not available")
            }
        }
    }

    @objc func echo(_ call: CAPPluginCall) {
        let value = call.getString("value") ?? ""
        call.resolve([
            "value": implementation.echo(value)
        ])
    }

    // MARK: - Private Properties
    private var pendingCall: CAPPluginCall?
}

// MARK: - UIDocumentPickerDelegate
extension AdvancedFileManagerPlugin: UIDocumentPickerDelegate {
    public func documentPicker(_ controller: UIDocumentPickerViewController, didPickDocumentsAt urls: [URL]) {
        guard let call = pendingCall else { return }

        var files: [[String: Any]] = []
        var directories: [[String: Any]] = []

        for url in urls {
            // 获取文件访问权限
            let accessing = url.startAccessingSecurityScopedResource()
            defer {
                if accessing {
                    url.stopAccessingSecurityScopedResource()
                }
            }

            do {
                let resourceValues = try url.resourceValues(forKeys: [
                    .fileSizeKey,
                    .contentModificationDateKey,
                    .creationDateKey,
                    .isDirectoryKey,
                    .contentTypeKey
                ])

                let isDirectory = resourceValues.isDirectory ?? false
                let fileSize = resourceValues.fileSize ?? 0
                let mtime = resourceValues.contentModificationDate?.timeIntervalSince1970 ?? 0
                let ctime = resourceValues.creationDate?.timeIntervalSince1970 ?? 0
                let contentType = resourceValues.contentType?.identifier ?? "application/octet-stream"

                let fileInfo: [String: Any] = [
                    "name": url.lastPathComponent,
                    "path": url.path,
                    "uri": url.absoluteString,
                    "size": fileSize,
                    "type": isDirectory ? "directory" : "file",
                    "mimeType": contentType,
                    "mtime": Int64(mtime * 1000),
                    "ctime": Int64(ctime * 1000)
                ]

                if isDirectory {
                    directories.append(fileInfo)
                } else {
                    files.append(fileInfo)
                }
            } catch {
                // 如果无法获取资源信息，创建基本信息
                let fileInfo: [String: Any] = [
                    "name": url.lastPathComponent,
                    "path": url.path,
                    "uri": url.absoluteString,
                    "size": 0,
                    "type": "file",
                    "mimeType": "application/octet-stream",
                    "mtime": Int64(Date().timeIntervalSince1970 * 1000),
                    "ctime": Int64(Date().timeIntervalSince1970 * 1000)
                ]
                files.append(fileInfo)
            }
        }

        call.resolve([
            "files": files,
            "directories": directories,
            "cancelled": false
        ])

        pendingCall = nil
    }

    public func documentPickerWasCancelled(_ controller: UIDocumentPickerViewController) {
        guard let call = pendingCall else { return }

        call.resolve([
            "files": [],
            "directories": [],
            "cancelled": true
        ])

        pendingCall = nil
    }
}

// MARK: - UIDocumentInteractionControllerDelegate
extension AdvancedFileManagerPlugin: UIDocumentInteractionControllerDelegate {
    public func documentInteractionControllerViewControllerForPreview(_ controller: UIDocumentInteractionController) -> UIViewController {
        return bridge?.viewController ?? UIViewController()
    }
}

File: package/android/src/main/java/com/mycompany/plugins/example/AdvancedFileManagerPlugin.java
package com.mycompany.plugins.example;

import android.Manifest;
import android.content.ContentResolver;
import android.content.Intent;
import android.content.pm.PackageManager;
import android.database.Cursor;
import android.net.Uri;
import android.os.Build;
import android.os.Environment;
import android.provider.DocumentsContract;
import android.provider.MediaStore;
import android.provider.OpenableColumns;
import android.provider.Settings;
import android.webkit.MimeTypeMap;
import androidx.activity.result.ActivityResult;
import androidx.core.app.ActivityCompat;
import androidx.core.content.ContextCompat;
import androidx.core.content.FileProvider;

import com.getcapacitor.JSArray;
import com.getcapacitor.JSObject;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.ActivityCallback;
import com.getcapacitor.annotation.CapacitorPlugin;
import com.getcapacitor.annotation.Permission;
import com.getcapacitor.annotation.PermissionCallback;

import java.io.File;
import java.io.FileOutputStream;
import java.io.InputStream;

@CapacitorPlugin(
    name = "AdvancedFileManager",
    permissions = {
        @Permission(
            strings = { Manifest.permission.READ_EXTERNAL_STORAGE, Manifest.permission.WRITE_EXTERNAL_STORAGE },
            alias = "storage"
        )
    }
)
public class AdvancedFileManagerPlugin extends Plugin {

    private AdvancedFileManager implementation;

    @Override
    public void load() {
        super.load();
        implementation = new AdvancedFileManager(getContext());
    }

    @PluginMethod
    public void requestPermissions(PluginCall call) {
        if (hasStoragePermissions()) {
            JSObject ret = new JSObject();
            ret.put("granted", true);
            ret.put("message", "Storage permissions already granted");
            call.resolve(ret);
        } else {
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.R) {
                // Android 11+ 需要特殊处理 MANAGE_EXTERNAL_STORAGE
                requestManageExternalStoragePermission(call);
            } else {
                // Android 10 及以下使用传统权限
                requestPermissionForAlias("storage", call, "storagePermissionCallback");
            }
        }
    }

    private void requestManageExternalStoragePermission(PluginCall call) {
        try {
            Intent intent = new Intent(Settings.ACTION_MANAGE_APP_ALL_FILES_ACCESS_PERMISSION);
            intent.setData(Uri.parse("package:" + getContext().getPackageName()));
            startActivityForResult(call, intent, "manageStoragePermissionResult");
        } catch (Exception e) {
            // 如果无法打开设置页面，尝试通用设置
            Intent intent = new Intent(Settings.ACTION_MANAGE_ALL_FILES_ACCESS_PERMISSION);
            startActivityForResult(call, intent, "manageStoragePermissionResult");
        }
    }

    @ActivityCallback
    private void manageStoragePermissionResult(PluginCall call, ActivityResult result) {
        JSObject ret = new JSObject();
        boolean granted = Environment.isExternalStorageManager();
        ret.put("granted", granted);
        ret.put("message", granted ? "MANAGE_EXTERNAL_STORAGE permission granted" : "MANAGE_EXTERNAL_STORAGE permission denied");
        call.resolve(ret);
    }

    @PluginMethod
    public void checkPermissions(PluginCall call) {
        JSObject ret = new JSObject();
        boolean granted = hasStoragePermissions();
        ret.put("granted", granted);
        ret.put("message", granted ? "Storage permissions granted" : "Storage permissions not granted");
        call.resolve(ret);
    }

    @PermissionCallback
    private void storagePermissionCallback(PluginCall call) {
        JSObject ret = new JSObject();
        boolean granted = hasStoragePermissions();
        ret.put("granted", granted);
        ret.put("message", granted ? "Storage permissions granted" : "Storage permissions denied");
        call.resolve(ret);
    }

    private boolean hasStoragePermissions() {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.R) {
            // Android 11+ 需要 MANAGE_EXTERNAL_STORAGE 权限
            return Environment.isExternalStorageManager();
        }

        return ContextCompat.checkSelfPermission(getContext(), Manifest.permission.READ_EXTERNAL_STORAGE) == PackageManager.PERMISSION_GRANTED &&
               ContextCompat.checkSelfPermission(getContext(), Manifest.permission.WRITE_EXTERNAL_STORAGE) == PackageManager.PERMISSION_GRANTED;
    }

    @PluginMethod
    public void listDirectory(PluginCall call) {
        String path = call.getString("path");
        Boolean showHidden = call.getBoolean("showHidden", false);
        String sortBy = call.getString("sortBy", "name");
        String sortOrder = call.getString("sortOrder", "asc");

        if (path == null) {
            call.reject("Path is required");
            return;
        }

        try {
            JSObject result = implementation.listDirectory(path, showHidden, sortBy, sortOrder);
            call.resolve(result);
        } catch (Exception e) {
            call.reject("Failed to list directory: " + e.getMessage());
        }
    }

    @PluginMethod
    public void createDirectory(PluginCall call) {
        String path = call.getString("path");
        Boolean recursive = call.getBoolean("recursive", false);

        if (path == null) {
            call.reject("Path is required");
            return;
        }

        try {
            implementation.createDirectory(path, recursive);
            call.resolve();
        } catch (Exception e) {
            call.reject("Failed to create directory: " + e.getMessage());
        }
    }

    @PluginMethod
    public void deleteDirectory(PluginCall call) {
        String path = call.getString("path");

        if (path == null) {
            call.reject("Path is required");
            return;
        }

        try {
            implementation.deleteDirectory(path);
            call.resolve();
        } catch (Exception e) {
            call.reject("Failed to delete directory: " + e.getMessage());
        }
    }

    @PluginMethod
    public void createFile(PluginCall call) {
        String path = call.getString("path");
        String content = call.getString("content", "");
        String encoding = call.getString("encoding", "utf8");

        if (path == null) {
            call.reject("Path is required");
            return;
        }

        try {
            implementation.createFile(path, content, encoding);
            call.resolve();
        } catch (Exception e) {
            call.reject("Failed to create file: " + e.getMessage());
        }
    }

    @PluginMethod
    public void readFile(PluginCall call) {
        String path = call.getString("path");
        String encoding = call.getString("encoding", "utf8");

        if (path == null) {
            call.reject("Path is required");
            return;
        }

        try {
            JSObject result = implementation.readFile(path, encoding);
            call.resolve(result);
        } catch (Exception e) {
            call.reject("Failed to read file: " + e.getMessage());
        }
    }

    @PluginMethod
    public void writeFile(PluginCall call) {
        String path = call.getString("path");
        String content = call.getString("content");
        String encoding = call.getString("encoding", "utf8");
        Boolean append = call.getBoolean("append", false);

        if (path == null || content == null) {
            call.reject("Path and content are required");
            return;
        }

        try {
            implementation.writeFile(path, content, encoding, append);
            call.resolve();
        } catch (Exception e) {
            call.reject("Failed to write file: " + e.getMessage());
        }
    }

    @PluginMethod
    public void deleteFile(PluginCall call) {
        String path = call.getString("path");

        if (path == null) {
            call.reject("Path is required");
            return;
        }

        try {
            implementation.deleteFile(path);
            call.resolve();
        } catch (Exception e) {
            call.reject("Failed to delete file: " + e.getMessage());
        }
    }

    @PluginMethod
    public void renameFile(PluginCall call) {
        String path = call.getString("path");
        String newName = call.getString("newName");

        if (path == null || newName == null) {
            call.reject("Path and newName are required");
            return;
        }

        try {
            implementation.renameFile(path, newName);
            call.resolve();
        } catch (Exception e) {
            call.reject("Failed to rename file: " + e.getMessage());
        }
    }

    @PluginMethod
    public void moveFile(PluginCall call) {
        String sourcePath = call.getString("sourcePath");
        String destinationPath = call.getString("destinationPath");

        if (sourcePath == null || destinationPath == null) {
            call.reject("sourcePath and destinationPath are required");
            return;
        }

        try {
            implementation.moveFile(sourcePath, destinationPath);
            call.resolve();
        } catch (Exception e) {
            call.reject("Failed to move file: " + e.getMessage());
        }
    }

    @PluginMethod
    public void copyFile(PluginCall call) {
        String sourcePath = call.getString("sourcePath");
        String destinationPath = call.getString("destinationPath");
        Boolean overwrite = call.getBoolean("overwrite", false);

        if (sourcePath == null || destinationPath == null) {
            call.reject("sourcePath and destinationPath are required");
            return;
        }

        try {
            implementation.copyFile(sourcePath, destinationPath, overwrite);
            call.resolve();
        } catch (Exception e) {
            call.reject("Failed to copy file: " + e.getMessage());
        }
    }

    @PluginMethod
    public void openSystemFilePicker(PluginCall call) {
        String type = call.getString("type", "file");
        Boolean multiple = call.getBoolean("multiple", false);
        JSArray acceptArray = call.getArray("accept");
        String title = call.getString("title", "选择文件");

        Intent intent;

        if ("directory".equals(type)) {
            // 选择目录
            intent = new Intent(Intent.ACTION_OPEN_DOCUMENT_TREE);
            intent.putExtra(DocumentsContract.EXTRA_PROMPT, title);
        } else {
            // 选择文件
            intent = new Intent(Intent.ACTION_OPEN_DOCUMENT);
            intent.addCategory(Intent.CATEGORY_OPENABLE);

            if (multiple) {
                intent.putExtra(Intent.EXTRA_ALLOW_MULTIPLE, true);
            }

            // 设置文件类型过滤
            if (acceptArray != null && acceptArray.length() > 0) {
                try {
                    String[] mimeTypes = new String[acceptArray.length()];
                    for (int i = 0; i < acceptArray.length(); i++) {
                        String ext = acceptArray.getString(i);
                        String mimeType = MimeTypeMap.getSingleton().getMimeTypeFromExtension(ext);
                        mimeTypes[i] = mimeType != null ? mimeType : "*/*";
                    }
                    intent.putExtra(Intent.EXTRA_MIME_TYPES, mimeTypes);
                } catch (Exception e) {
                    intent.setType("*/*");
                }
            } else {
                intent.setType("*/*");
            }
        }

        startActivityForResult(call, intent, "filePickerResult");
    }

    @ActivityCallback
    private void filePickerResult(PluginCall call, ActivityResult result) {
        JSObject ret = new JSObject();
        JSArray files = new JSArray();
        JSArray directories = new JSArray();

        if (result.getResultCode() == getActivity().RESULT_OK && result.getData() != null) {
            Intent data = result.getData();
            String requestType = call.getString("type", "file");

            if (data.getClipData() != null) {
                // 多选
                for (int i = 0; i < data.getClipData().getItemCount(); i++) {
                    Uri uri = data.getClipData().getItemAt(i).getUri();
                    JSObject fileInfo = getFileInfoFromUri(uri);
                    if (fileInfo != null) {
                        if ("directory".equals(fileInfo.getString("type"))) {
                            directories.put(fileInfo);
                        } else {
                            files.put(fileInfo);
                        }
                    }
                }
            } else if (data.getData() != null) {
                // 单选
                Uri uri = data.getData();
                JSObject fileInfo = getFileInfoFromUri(uri);
                if (fileInfo != null) {
                    if ("directory".equals(requestType) || "directory".equals(fileInfo.getString("type"))) {
                        directories.put(fileInfo);
                    } else {
                        files.put(fileInfo);
                    }
                }
            }

            ret.put("files", files);
            ret.put("directories", directories);
            ret.put("cancelled", false);
        } else {
            ret.put("files", files);
            ret.put("directories", directories);
            ret.put("cancelled", true);
        }

        call.resolve(ret);
    }

    @PluginMethod
    public void openSystemFileManager(PluginCall call) {
        String path = call.getString("path");

        try {
            Intent intent = new Intent(Intent.ACTION_VIEW);

            if (path != null && !path.isEmpty()) {
                File file = new File(path);
                if (file.exists()) {
                    Uri uri;
                    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N) {
                        uri = FileProvider.getUriForFile(getContext(),
                            getContext().getPackageName() + ".fileprovider", file);
                    } else {
                        uri = Uri.fromFile(file);
                    }
                    intent.setDataAndType(uri, "resource/folder");
                    intent.addFlags(Intent.FLAG_GRANT_READ_URI_PERMISSION);
                }
            } else {
                // 打开默认文件管理器
                intent = getContext().getPackageManager().getLaunchIntentForPackage("com.android.documentsui");
                if (intent == null) {
                    // 备用方案
                    intent = new Intent(Intent.ACTION_VIEW);
                    intent.setType("resource/folder");
                }
            }

            getActivity().startActivity(intent);
            call.resolve();
        } catch (Exception e) {
            call.reject("Failed to open file manager: " + e.getMessage());
        }
    }

    @PluginMethod
    public void openFileWithSystemApp(PluginCall call) {
        String filePath = call.getString("filePath");
        String mimeType = call.getString("mimeType");

        if (filePath == null) {
            call.reject("File path is required");
            return;
        }

        try {
            File file = new File(filePath);
            if (!file.exists()) {
                call.reject("File does not exist: " + filePath);
                return;
            }

            Uri uri;
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N) {
                uri = FileProvider.getUriForFile(getContext(),
                    getContext().getPackageName() + ".fileprovider", file);
            } else {
                uri = Uri.fromFile(file);
            }

            Intent intent = new Intent(Intent.ACTION_VIEW);

            if (mimeType == null || mimeType.isEmpty()) {
                // 自动检测 MIME 类型
                String extension = getFileExtension(filePath);
                mimeType = MimeTypeMap.getSingleton().getMimeTypeFromExtension(extension);
                if (mimeType == null) {
                    mimeType = "*/*";
                }
            }

            intent.setDataAndType(uri, mimeType);
            intent.addFlags(Intent.FLAG_GRANT_READ_URI_PERMISSION);
            intent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK);

            getActivity().startActivity(Intent.createChooser(intent, "打开文件"));
            call.resolve();
        } catch (Exception e) {
            call.reject("Failed to open file: " + e.getMessage());
        }
    }

    private JSObject getFileInfoFromUri(Uri uri) {
        JSObject fileInfo = new JSObject();
        ContentResolver contentResolver = getContext().getContentResolver();

        try {
            // 获取文件基本信息
            Cursor cursor = contentResolver.query(uri, null, null, null, null);
            if (cursor != null && cursor.moveToFirst()) {
                // 获取文件名
                int nameIndex = cursor.getColumnIndex(OpenableColumns.DISPLAY_NAME);
                String fileName = nameIndex >= 0 ? cursor.getString(nameIndex) : "unknown";

                // 获取文件大小
                int sizeIndex = cursor.getColumnIndex(OpenableColumns.SIZE);
                long fileSize = sizeIndex >= 0 ? cursor.getLong(sizeIndex) : 0;

                // 获取 MIME 类型
                String mimeType = contentResolver.getType(uri);

                // 判断是否为目录
                boolean isDirectory = DocumentsContract.isDocumentUri(getContext(), uri) &&
                                    DocumentsContract.Document.MIME_TYPE_DIR.equals(mimeType);

                // 尝试获取真实路径
                String realPath = getRealPathFromUri(uri);

                fileInfo.put("name", fileName);
                fileInfo.put("path", realPath != null ? realPath : uri.toString());
                fileInfo.put("uri", uri.toString());
                fileInfo.put("size", fileSize);
                fileInfo.put("type", isDirectory ? "directory" : "file");
                fileInfo.put("mimeType", mimeType != null ? mimeType : "application/octet-stream");
                fileInfo.put("mtime", System.currentTimeMillis());
                fileInfo.put("ctime", System.currentTimeMillis());

                cursor.close();
                return fileInfo;
            }
            if (cursor != null) {
                cursor.close();
            }
        } catch (Exception e) {
            // 如果查询失败，返回基本信息
            fileInfo.put("name", "unknown");
            fileInfo.put("path", uri.toString());
            fileInfo.put("uri", uri.toString());
            fileInfo.put("size", 0);
            fileInfo.put("type", "file");
            fileInfo.put("mimeType", "application/octet-stream");
            fileInfo.put("mtime", System.currentTimeMillis());
            fileInfo.put("ctime", System.currentTimeMillis());
        }

        return fileInfo;
    }

    private String getRealPathFromUri(Uri uri) {
        try {
            // 对于 content:// URI，尝试获取真实路径
            if ("content".equals(uri.getScheme())) {
                if (DocumentsContract.isDocumentUri(getContext(), uri)) {
                    // 处理 Documents Provider
                    String docId = DocumentsContract.getDocumentId(uri);

                    if ("com.android.externalstorage.documents".equals(uri.getAuthority())) {
                        // 外部存储 - 这是最常见的情况
                        return handleExternalStorageDocument(docId);
                    } else if ("com.android.providers.downloads.documents".equals(uri.getAuthority())) {
                        // 下载目录
                        return handleDownloadsDocument(docId);
                    } else if ("com.android.providers.media.documents".equals(uri.getAuthority())) {
                        // 媒体文件
                        return handleMediaDocument(docId);
                    }
                } else {
                    // 普通 content URI
                    return getDataColumn(uri, null, null);
                }
            } else if ("file".equals(uri.getScheme())) {
                // 文件 URI
                return uri.getPath();
            }
        } catch (Exception e) {
            // 忽略错误，返回 null
        }

        return null;
    }

    private String handleExternalStorageDocument(String docId) {
        String[] split = docId.split(":");
        if (split.length >= 2) {
            String type = split[0];
            String path = split[1];

            if ("primary".equalsIgnoreCase(type)) {
                // 主要存储 (内部存储)
                if (path.isEmpty()) {
                    return "/storage/emulated/0";
                } else {
                    return "/storage/emulated/0/" + path;
                }
            } else {
                // 外部 SD 卡或其他存储
                // 尝试一些常见的路径
                String[] possiblePaths = {
                    "/storage/" + type + "/" + path,
                    "/mnt/media_rw/" + type + "/" + path,
                    "/storage/sdcard1/" + path
                };

                for (String possiblePath : possiblePaths) {
                    File file = new File(possiblePath);
                    if (file.exists()) {
                        return possiblePath;
                    }
                }

                // 如果都不存在，返回最可能的路径
                return "/storage/" + type + "/" + path;
            }
        }
        return null;
    }

    private String handleDownloadsDocument(String docId) {
        if (docId.startsWith("raw:")) {
            return docId.replaceFirst("raw:", "");
        }

        // 对于数字 ID，尝试查询下载数据库
        try {
            Uri contentUri = Uri.parse("content://downloads/public_downloads");
            Uri uri = Uri.withAppendedPath(contentUri, docId);
            return getDataColumn(uri, null, null);
        } catch (Exception e) {
            // 如果查询失败，返回默认下载路径
            return "/storage/emulated/0/Download";
        }
    }

    private String handleMediaDocument(String docId) {
        String[] split = docId.split(":");
        if (split.length >= 2) {
            String type = split[0];
            String id = split[1];

            Uri contentUri = null;
            if ("image".equals(type)) {
                contentUri = MediaStore.Images.Media.EXTERNAL_CONTENT_URI;
            } else if ("video".equals(type)) {
                contentUri = MediaStore.Video.Media.EXTERNAL_CONTENT_URI;
            } else if ("audio".equals(type)) {
                contentUri = MediaStore.Audio.Media.EXTERNAL_CONTENT_URI;
            }

            if (contentUri != null) {
                String selection = "_id=?";
                String[] selectionArgs = new String[]{id};
                return getDataColumn(contentUri, selection, selectionArgs);
            }
        }
        return null;
    }

    private String getDataColumn(Uri uri, String selection, String[] selectionArgs) {
        Cursor cursor = null;
        final String column = "_data";
        final String[] projection = {column};

        try {
            cursor = getContext().getContentResolver().query(uri, projection, selection, selectionArgs, null);
            if (cursor != null && cursor.moveToFirst()) {
                final int columnIndex = cursor.getColumnIndexOrThrow(column);
                return cursor.getString(columnIndex);
            }
        } catch (Exception e) {
            // 忽略错误
        } finally {
            if (cursor != null) {
                cursor.close();
            }
        }
        return null;
    }

    private String getFileExtension(String filePath) {
        int lastDot = filePath.lastIndexOf('.');
        if (lastDot > 0 && lastDot < filePath.length() - 1) {
            return filePath.substring(lastDot + 1).toLowerCase();
        }
        return "";
    }

    @PluginMethod
    public void echo(PluginCall call) {
        String value = call.getString("value");

        JSObject ret = new JSObject();
        ret.put("value", implementation.echo(value));
        call.resolve(ret);
    }
}
